{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "366435e7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "EvolTree running in new folder: /Users/jra/Dropbox/ciliates/P0460249_riboprotein/P0460249_riboprotein_ohnologs.fa.aligned.trimal.automated1.cds.evolTree\n",
      "\n",
      "      /-5888.PTET_51_T0460249\n",
      "   /-|\n",
      "  |   \\-5888.PTET_51_T1750015\n",
      "--|\n",
      "  |--5888.PTET_51_T0360256\n",
      "  |\n",
      "   \\-5888.PTET_51_T0210034\n"
     ]
    }
   ],
   "source": [
    "from ete3 import EvolTree\n",
    "from ete3.treeview.layouts import evol_clean_layout\n",
    "from ete3 import NCBITaxa\n",
    "import matplotlib.pyplot as plt\n",
    "import os\n",
    "import shutil\n",
    "ncbi = NCBITaxa()\n",
    "\n",
    "def file_exists(filename):\n",
    "    current_directory = os.getcwd()\n",
    "    if os.path.exists(filename):\n",
    "        print(f\"The file {filename} exists in the current working directory {current_directory}.\")\n",
    "    else:\n",
    "        print(f\"The file {filename} does not exist in the current working directory {current_directory}.\")\n",
    "\n",
    "def organize_your_life(input_alignment):\n",
    "    # get the full path of the input file\n",
    "    full_input_file_path = os.path.abspath(input_alignment)\n",
    "    full_input_newick_path = full_input_file_path + '.nwk'\n",
    "    \n",
    "    directory, alignment = os.path.split(full_input_file_path)\n",
    "    directory, newick = os.path.split(full_input_newick_path)\n",
    "\n",
    "    path = os.path.join(directory, f\"{alignment}.evolTree\")\n",
    "    os.makedirs(path, exist_ok=True)\n",
    "    \n",
    "    '''\n",
    "    #still learning\n",
    "    print(f\"directory: {directory}\")\n",
    "    print(f\"new_directory: {path}\")\n",
    "    print(f\"input_alignment: {input_alignment}\")\n",
    "    print(f\"input_file_path: {full_input_file_path}\")\n",
    "    print(f\"input_alignmentname: {input_alignmentname}\")\n",
    "    print(f\"input_newick: {input_newick}\")\n",
    "    print(f\"input_newickname: {input_newickname}\")\n",
    "    '''\n",
    "    \n",
    "    # copy input files the new directory\n",
    "    shutil.copy(full_input_file_path, path)\n",
    "    shutil.copy(full_input_newick_path, path)\n",
    "\n",
    "    # move to the new directory\n",
    "    os.chdir(path)\n",
    "    print(f\"EvolTree running in new folder: {path}\")\n",
    "    \n",
    "    # load the tree\n",
    "    tree = EvolTree(newick = full_input_newick_path, format=1)\n",
    "    # associate alignment to tree\n",
    "    tree.link_to_alignment(input_alignment)\n",
    "    tree.workdir = path # set working directory where models will be saved\n",
    "    alnPrefix = alignment # might need this in the future, like for saving files?\n",
    "    return tree, alnPrefix\n",
    "    \n",
    "# Get species name from TaxID\n",
    "def get_species_name(taxid):\n",
    "    try:\n",
    "        lineage = ncbi.get_lineage(taxid)\n",
    "        names = ncbi.get_taxid_translator(lineage)\n",
    "        species_name = names[taxid]\n",
    "        return species_name\n",
    "    except:\n",
    "        return f\"Unknown_species_{taxid}\"\n",
    "\n",
    "tree, alnPrefix = organize_your_life('/Users/jra/Dropbox/ciliates/P0460249_riboprotein/P0460249_riboprotein_ohnologs.fa.aligned.trimal.automated1.cds')\n",
    "\n",
    "#print(alignment)\n",
    "#print(newick)\n",
    "#print(os.getcwd())\n",
    "print(tree)\n",
    "#print(prefix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "28553896",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5888.PTET_51_T0460249\tParamecium tetraurelia\t0.1388\n",
      "5888.PTET_51_T1750015\tParamecium tetraurelia\t0.1398\n",
      "5888.PTET_51_T0360256\tParamecium tetraurelia\t0.7080\n",
      "5888.PTET_51_T0210034\tParamecium tetraurelia\t0.0000\n",
      "Ds values have been written to P0460249_riboprotein_ohnologs.fa.aligned.trimal.automated1.cds.ref.dS.txt\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAArgAAAGDCAYAAADAlWPcAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAsTAAALEwEAmpwYAAAopklEQVR4nO3debhcRZnH8d8PwhJkNUEGISYuBAGFCEHFNYobCCKCuEAkjiPL6CgOiMigRsF1VNBhkE2NIANkFBQYRtYJooBCMAQBg6BgggRI2Dch4Z0/qpqcbrvv7WzdnTrfz/Oc53af9T1Ln3q7uk5dR4QAAACAUqzW7wAAAACAFYkEFwAAAEUhwQUAAEBRSHABAABQFBJcAAAAFIUEFwAAAEUhwQWWgu1pto/pw3bD9kt6vd1u2Z5ke95yLH+i7c+tyJj6wfaWtn9n+xHbn+h3PMvC9gzb/7QC1mPbP7T9gO3frojYsHxsv9b2H20/avvd/Y4HWJlIcIHlYPtI23/OBcY822d3mO8i219qM34P2/Ntj1j50Q4G21Ns/6o6LiIOioijV8K27rD9lhW93iEcLmlGRKwXEd/t4XYH0eskvVXS5hHxyuFmzl8en8qfpcawemX6m21fb/th23+yfUBlmm0fY/su2w/lJH2byvRxti/MyfZ828d3+szZnmr76bz9B21fZXun/FlvxPWk7cWV9zflZcP2Yy37cHj+Atd4/1Rl/Y/a/t9hjkvrOk+tTHtZvrcssN1Np/ZfknR8RKwbET/rYv5VQj7fT+bjs8D2ObY3zdNs++u2F+bhG7bdZh1vzMf6mJbxG9v+r3wtPGD7jMq0tWz/IF+T823/a2XaeNs/t32f7fvzedqyQ/yX522PqIzbKo9/yPZttvdcEceqTkhwgWVke39JkyW9JSLWlTRR0mUdZp8maXKbG+tkSWdExKKVFugKUKcEvJMuj8FYSTetxPWvSsZKuiMiHluKZb6Rk6/GsFiSbK8h6VxJJ0naQNL7JH3b9nZ5ufdK+kdJr5f0XElXSzq9st4TJN0raVNJEyS9UdI/DxHH2fkzvbGkX0k6R9JXG3FJOkjS1ZU4t6ksu13LPnwjf4FrLPuVxvrzsEsXx6W6zmrt+tOSpkv6SBfrkMq+Pj+ej+94SRtKOjaPP0DSuyVtJ2lbSbtJOrC6YL6+viPpN23We46k+UrH7nmSvlmZNlXSFnnamyQdbvsdedqGks6TtKWkTST9VtLPW1due19JI1rGjcjzXqB0PR8g6ce2xw+x/2hBggsMwfYrcq3RI061s2tXJu8o6aKIuF2SImJ+RJzcYVU/U7pRvb6y7o2Ubran2X6l7atzLcHduYZpzQ4xNf2E7JYaUdsvtX1JrjWYY3ufyrRdbd+c9+cu24d12MYU27+2fazt+yVNzbUV37T9F9v35FqpkR2WP8L27Xk7NzdqH2xvJelESTvl2pYH8/hnm37YvsX2bpV1jci1Mtvn9692qlV70PYNtid1iOF0SS+QdH7e1uF5/Lts35SXn5FjaivXqnzM9h8l/TGP2832LC+p3ds2j79cqZA7Pm9v/FDHzLlZh+3P2J4v6Ye2V6scu4W2p9t+bp5/XI5n/7y+Bbb/rRLr6k61jI3jPtP2mOGuieHY/sd8Th5wqoUaW5n2HdtznWqwZtp+fR7/EUmnVs7zF7vdXgfPlbS+pNMjuVbSLZK2ztNfKOlXEfGnnBT/uDKtMX16RDwZEfMl/UJSNSltKyKelvQjSf8gadRy7sMKFxFzIuL76iJptX27pBdpyedhLdvPt31evi5us/3RyvxTbf/E9o9tPyxpSpt1jrJ9fj7/1zrVond7L5pm+z9t/0++Xn9j+8XdLDvMMblf0k8lvSyP2l/StyJiXkTcJelbbfblUEkXS/pDy/69TdIYSZ+OiIci4umI+F1llg9JOjoiHoiIWySd0lh3RPw2Ir4fEffn6+hYSVvaHlVZ/waSvqD0y0/VSyU9X9KxEbE4Ii6X9GulChF0KyIYGBjaDJLWlHSnpE9JWkPS3ko1Jsfk6ftJul/Sp5Vqb1cfZn2nSDq18v5ASbPy6x0kvVrpm/w4pcL7kMq8Iekl+fUMSf9UmTZFqXCXpOdImivpw3ld20taIGmbPP1uSa/PrzeStH2HWKdIWiTpX/J6Rko6TqlG4rmS1pN0vlKtliRNkjSvsvx7lW7QqynVtj0madPWeCvzT6sc188r1Wo3pr1T0h/y680kLZS0a173W/P7jTvsxx1KNeyN9+NzLG/N5/RwSbdJWrPD8iHpkrzPI/PxvFfSqyStrlR43iFprQ7nZrhjtkjS1yWtldd/iKRrJG2ex50k6cw8/7gczyl53u0k/U3SVnn6pyXdqFRj5Dx91HDXRJt9fnYflGq+bpO0VV72KElXVebdL29jhFKSMF/S2u3Os6QPSpo9xOdjmtLn6X5JMyXt1TL9vyR9LB/3nfJ5GJOnjZV0fT6/a0j6hqSfVZY9SNJpktZRuoZ+L2nPDnFMlfTj/HotSf8uaW6bz8ev2iz77Od0iP18dv1d3odC0l/zsT1H0rg287xEUnSxrjvU/Hm4Qql2e22lmu37JO1cifPpfA2sJmlkm/WdlYd1lL5QzFX396LG+X5lnn6GpLO6WXaYa3a0pMuVvgxJ0kOSXlWZd6KkRyrvx0q6VdK6qtyHKveii5S+MC2UdK2kN1bunyFpk8r8e0u6sUOM75Z0d8u4/1QqX8bldY3I418u6VFJrsx7iaRzu71uGIIEl4Gh0yDpDblgqd5krmq5Ae4r6VKlpGmhpCOGWN/r8s12ZH7/a0mf6jDvIdWbmbpPcN8n6cqWdZ0k6Qv59V+UEuv1h9n3KZL+UnnvvI8vrozbSdKf8+tJqiS4bdY3S9IerfFWpj9bsCgV1o9IWie/P0PS5/PrzzQKrsqyF0nav8N271Bzgf45pZq8xvvVJN0laVKH5UPSmyvvv6dUY1OdZ46WFHrPnpsuj9lTyglhHneLcoKR32+qlGQ0vviEUpvWxvTfSnp/JY492uzDkNdEm/mr+/C/kj7ScrwelzS2w7IPKP2c3vY8D3PNba8lyfKu+Rp4bWX67pLuUfpSsEjSRyvT1lT6iTnytD9LemFl+lZKSfOiPM80VT7XLXFMzeflQaUk+nJJO7T5fHRKcB/OyzaGt7dZ/9IkuG/I+7ehpOOVkvMRLfMsdYKrVDO5WNJ6lelflTStEucvh1jX6vna3LIy7hh1fy+apuYv/LtqyRfZZblmH8/H+y6le8bGedpiSS+tzLtFPk/O738u6X2VmKr395PzvB9R+uL0/ryN0fn4hZo/v29VapbTGt/mOa4PVMZNVLovVj/bjQR3DUl/UvoCvoaktyldkxd1e90wBE0UgCE8X9Jdke842Z3VGSLijIh4i1Lhc5CkL9l+e7uVRcSvlGpI9rD9IqUmDv8lPftAwgVODyo8rNROb/QyxDxW0qvyz+cP5iYA+yr9xCpJeykVJHfavsL2TkOsa27l9cZKtTQzK+v9RR7/d2x/qPIz/oNKPxd2tT8RcZtSore77XUkvUv5OOX9e2/L/r1OKRHsxvNVOYcR8Uzez82GWKZ6HMZKOrRl+2Pyelt1c8zui4gnW9Z/bmX+W5QK6E0q88yvvH5cqeZJOY7b28Qx3DUxlLGSvlNZ7n6lxH0zSbJ9aG6+8FCevoGW7bpVRFwfEQsjYlFEXKiUpLwnb+elks5W+kl4TaXmBYfbfmde/AtKn6cxSrWRX5R0ue11bK+m9CXoHKWawdFKtW9fHyKc6RGxYUQ8LyLeHBEzl2JXts/LNoaLlmLZvxMRv4yIpyLiQUmfVGpu0bFZzVJ4vqT7I+KRyrg71fxZmKvONlZKzqrztH5WhrvuOl3Ly3LNfiIf780iYt+IuC+Pf1SpeUvD+pIejYiwvbtSgt/24WBJTyglrN+P1DzhrLyPr83rbayvuu7q8ZTtjZWaP5wQEWfmcasp1Zx/Mto8fxGpScO7lX69mq/068h0ScvcU00dkeACnd0taTO76cGwF7SbMd/8/lvSbC1p+9XOaUqF9GRJF0fEPXn895Taf20REetLOlIpkWjnMaXEqaF6058r6YqWAnbdiDg4x3ltROyh9LDEz5Rump1UE/sFSjf7bSrr3SDSQx1NchvNUyR9XNKoiNhQqdapsT/RukwbZ0r6gKQ9JN2ck97G/p3esn/PiYivdbEPUqqRH1uJ1UpJ0V1DxFJdx1xJX27Z/jqNgqtFN8esNb65knZpWf/akdoODmeupBd3GN/xmuhinQe2LDsyIq7K7W0/I2kfSRvl8/yQOl+3Sysq63qZpDkRcVFEPBMRcyT9j6TGA1rbKT24NS8nyNOUktitlZqHjFHqPeBvEbFQ0g+VvuitiqrHZXn8VdJzba9XGfcCNX8Whvqs3qdUI755ZdyYyuvlve6WddlWNyldHw3baUmb5Z0lTcwVC/OVao4Psd14GGy2OhyDiHhAqYzotO7GcxYXSzovIr5cmW99pRrcs/N2r83j5zXasUfE7Ih4Y0SMioi3K7Wfpru9pUCCC3R2tdIN/BNODzq9R6m9mKRnH8R6p+31nB4O2kWpZqndk7gNp0l6i6SPKj280rCe0k+bj+baqqFu5LMkvSfXTr1EzU9QXyBpvO3JttfIw45OXc6saXtf2xvkGoKHlWoHh5VrOk+RdKzt5+X936xDbfVzlAqF+/J8H1Zz0n+PpM3d4SG67Cyln+UO1pLaWym1hdvd9tudHqpa2+lhrc3briVt60WV99MlvdP2zk5PTh+q1I71qiFiqTpF0kG2X+XkOY1roHXGpTxmDSdK+nL+ktDoomiPLmM7VdLRtrfIsW2bH2jpeE10sc4TJX3Wucst2xvYfm+etp7S5+M+SSNsf17NtVlLxfbettfNn6W3KbXvPS9P/p2kLZy6CrPTw0i7SbohT79WqWZ/k7z8ZKWfdm+LiAVKTRYOzp/jDZXaTt+gAWd7G9sT8rW+rtIDUncp1ew3usBaW6lWW/nzsFY3646IuUrX/Vfzctsq3UvOGHrJZ5dfrFQrPjXfi16q9OW9YXmuu+VZttVpkv41f/aer/SZn5anfU6p3faEPJyn9Jn9cJ5+rqSNnB7sXN323ko13L+urPso2xvl/f9oY92211f65eDXEXFES0wPKdWgN7bb+LK1g3L5kT+/a+dje5jSr1TThK6R4AIdRMRTSj+RTlFqW/g+pRt6w8NKNa1/UWqX9Q1JB+emCJ3WeYdSofIcLSm8JekwpYdwHlG6wXb6yUxKT+M+pZS8/UiVAin/3Pg2pbZijQdTGg8xSanm+A6nZhAHKSUR3fqM0gNH1+TlL1V6oKl1H29WKoivzjG+XEsKBCm1abxJ0nzbC9ptKCLuzsu/RpVjkQvlPZSO+31KNT2fVud72VeVCqAHbR+Wa/72k/QfSjWsu0vaPZ/rYUXEdUqF2PFK18RtavN0eUVXx6ziO0rXxcW2H1F64OxV3cQm6dtKCfzFStfm95Xaew93TXQUEefmec/K8f9eS2pNL1Jqo3ur0k/bT2qIn7Tzl6uhnvb/pFLy9qDSg10fjYgZOY7blboB+27etyuUnpT/fl7260oJ66y8/KeUHlJ7ME9/j6R3KF0ztykl5p/Kcb3AqVeBtr/OLIMb3NwP7nHLsa5NlK7/h5XaZI6TtFv+giqlXyOe0JJawyeU2mJ36wN5nX9VSua+EBGXLMXyH1dqljJfqVu2M5W+MHZzL+poeZZt4ySlhztvVLp+/yePU0Q8Eqn3m/mRetd4QtJjkXpiUP77LqX780OSjlBq5964b31BqVnQnUrX5L9HxC/ytD2Vms18uOV6eEEk1e02mlPcU7kXTVaqIb5Xqab5rRHxt2XY/9pqNLIGAABYZra/LukfImL/fscCUIMLAACWmlNftdvmphKvVGricG6/4wIkElwAAPomN914tM2wTP9xrMfWU2q29ZhS85hvqc1/6wL6gSYKAAAAKAo1uAAAACgKCS4AAACKMqLfAWCwjB49OsaNG9fvMAAAwMrwzDPp72pl1HHOnDlzQUT83X/VJMFFk3Hjxum6667rdxgAAADDsn1nu/FlpO8AAAAY3gknpKFwJLgAAAB1MX16GgpHggsAAICikOACAACgKCS4AAAAKAoJLgAAAIpCN2EAAAB1MWNGvyPoCWpwAQAAUBQSXAAAgLr45jfTUDgSXAAAgLq44II0FI4EFwAAAEUhwQUAAEBR6EUBTWbOnCnb/Q4DAFAREf0OAVilkOACAADUxciR/Y6gJ8y3QlTZ5oIAgAFDWQ20Z3tmRExsHU8bXAAAABSFBBcAAKAujj46DYUjwQUAAKiLyy5LQ+FIcAEAAFAUElwAAAAUhQQXAAAARaEfXAAAgLoYNarfEfQE/eCiCf3gAsDgoawG2qMfXAAAANQCCS4AAEBdfPazaSgcbXABAADq4uqr+x1BT1CDCwAAgKKQ4AIAAKAoJLgAAAAoCm1wAQAA6mLzzfsdQU/QDy6a0A8uAAweymqgPfrBBQAAQC2Q4AIAANTFIYekoXC0wQUAAKiLWbP6HUFPUIMLAACAopDgAgAAoCgkuAAAACgKbXABAADqYvz4fkfQE/SDiyb0gwsAg4eyGmiPfnABAABQCyS4AAAAdXHAAWkoHG1wAQAA6uLWW/sdQU9QgwsAAICikOACAACgKCS4AAAAKAptcAEAAOpiwoR+R9AT9IOLJvSDCwCDh7IaaI9+cAEAAFALJLgAAAB1sd9+aSgcbXABAADqYt68fkfQE9TgAgAAoCgkuAAAACgKCS4AAACKQhtcAACAuthpp35H0BP0g4sm9IMLAIOHshpoj35wAQAAUAskuAAAAHWx115pKBxtcAEAAOpi4cJ+R9AT1OACAACgKCS4AAAAKAoJLgAAAIpCG1wAAIC62HnnfkfQE/SDiyb0gwsAg4eyGmiPfnABAABQCyS4AAAAdbHLLmkoHG1wAQAA6uKJJ/odQU9QgwsAAICikOACAACgKCS4AAAAKAptcAEAAOpit936HUFP0A8umtAPLgAMHspqoD36wQUAAEAtkOACAADUxaRJaSgcCS4AAACKQoILAACAopDgAgAAoCgkuAAAACgK/eACAADUxT779DuCnqAfXDShH1wAGDyU1UB79IMLAABQd48/nobC0UQBAACgLnbdNf2dMaOvYaxs1OACAACgKCS4AAAAKAoJLgAAAIpCggsAAICi8JAZAABAXUyZ0u8IeoJ+cNGEfnABYPBQVgPt0Q8uAABA3S1YkIbC0UQBAACgLvbeO/2lH1wAAABg1UGCCwAAgKKQ4AIAAKAoJLgAAAAoCg+ZAQAA1MXBB/c7gp6gH1w0oR9cABg8lNVAe/SDCwAAUHdz56ahcDRRAAAAqIvJk9Nf+sEFAAAAVh0kuAAAACgKCS4AAACKQoILAACAovCQGQAAQF0cemi/I+gJ+sFFE/rBBYDBQ1kNtEc/uAAAAHU3Z04aCkcTBQAAgLo48MD0l35wAQAAgFUHCS4AAACKQoILAACAopDgAgAAoCg8ZAYAAFAXRx3V7wh6gn5w0YR+cAFg8FBWA+3RDy4AAEDdzZqVhsLRRAEAAKAuDjkk/aUfXAAAAGDVQYILAACAopDgAgAAoCgkuAAAACgKD5kBAADUxVe+0u8IeoIEFwAAoC5e85p+R9ATNFEAAACoi6uuSkPhqMEFAACoiyOPTH/pBxcAAABYdZDgAgAAoCgkuAAAACgKCS4AAACKwkNmAAAAdXHccf2OoCdIcAEAAOpiwoR+R9ATNFEAAACoi0svTUPhHBH9jgEDxDYXBAAMGMpqrDCTJqW/hfSDa3tmRExsHU8NLgAAAIpCggsAAICikOACAACgKCS4AAAAKArdhAEAANTFSSf1O4KeIMEFAACoiy237HcEPUETBQAAgLo4//w0FI5+cNGEfnABYPBQVmOFoR9cAAAAYNVDggsAAICikOACAACgKCS4AAAAKArdhAEAANTF6af3O4KeIMEFAACoizFj+h1BT9BEAQAAoC7OPjsNhaMfXDShH1wAGDyU1Vhh6AcXAAAAWPWQ4AIAAKAoJLgAAAAoCgkuAAAAikI3YQAAAHXxk5/0O4KeIMEFAACoi9Gj+x1BT9BEAQAAoC6mTUtD4egHF03oBxcABg9lNVYY+sEFAAAAVj0kuAAAACgKCS4AAACKQoILAACAogyb4NpebHuW7d/b/m/b6/QisGVh+0LbG66E9R65Iudb0WxPsX18fn2Q7Q/1Iw4AADDgLrwwDYUbthcF249GxLr59RmSZkbEt4ddsT0iIhatmDD7q3oMlmU+21Y61s8sRwwdj6ftKZImRsTHl3X9lXXxqC4ADBh6UQDaW1G9KFwp6SW2d7f9G9u/s32p7U3yRqbaPtn2xZJOsz3O9pW2r8/Da/J8k2xfYXu67Vttf832vrZ/a/tG2y/O821s+6e2r83Da/P4dW3/MM872/Zeefwdtkfn7f6+svOH2Z6aX8+wfaztX9q+xfaOts+x/Ufbx7Q5cF+TNDLXYp+Rx+2XY51l+yTbq7fOl2O4xfYJkq6XNMb292xfZ/sm21+sbOMO26Pz64m2Z3Q4nm2PR0u8U20fll9/NM93Q15uYGvfAQBAD5xwQhoK13WCa3uEpF0k3SjpV5JeHRGvkHSWpMMrs+4gaY+I+KCkeyW9NSK2l/Q+Sd+tzLedpE9KermkyZLGR8QrJZ0q6V/yPN+RdGxE7ChprzxNkj4n6aGIeHlEbCvp8u53WZL0VES8QdKJkn4u6WOSXiZpiu1R1Rkj4ghJT0TEhIjY1/ZWeV9eGxETJC2WtG/rfHnxLSWdFhGviIg7Jf1b/paxraQ32t62i1irx7PT8ejknIjYMSK2k3SLpI90sT0AAFCq6dPTULhu/lXvSNuz8usrJX1fKXE72/amktaU9OfK/OdFxBP59RqSjrc9QSkRHF+Z79qIuFuSbN8u6eI8/kZJb8qv3yJp6/QLvyRpfdvr5fHvb4yMiAe62I+q8yrbuqkSx58kjZG0cIhld1ZKOq/NcY1USuTbuTMirqm838f2AUrHfVNJW0uaPVyslePZ6Xh08rJcK72hpHUlXdRuphzTAcPEAQAAsEroJsF9ItdUPsv2f0j6dkScZ3uSpKmVyY9VXn9K0j1KtbWrSXqyMu1vldfPVN4/U4lrNUk7VRK8xvYtaagGSYvUXDu9dsv06rZa4xjumFjSjyLis8PMJ1WOhe0XSjpM0o4R8YDtaZW4qvG2xlo9np2OR6ftT5P07oi4IbfTndRupog4WdLJeV009AIAAKu0Ze0mbANJd+XX+w8z39354arJklZfyu1cLOnZB6dyTXC78Ru1LHePpOfZHmV7LUm7LeV2Wz1te438+jJJe9t+Xt72c22PbTNfq/WVktWHcpvlXSrT7lCqFZZS04NOOh2PTtaTdHeOad9h5gUAACjCsia4UyX9t+0rJS0YYr4TJO1v+xql5gmPDTFvO5+QNDE/SHazpIPy+GMkbeTUddkNWtKkQZIUEU9L+pKk30i6QNIflnK7rU6WNNv2GRFxs6SjJF1se7akS5SaGzTN17qCiLhB0u8k3STpB5J+XZn8RUnfycdz8RBxdDoenXxO6RhcouU/BgAAAKuEYbsJQ73QRAEABg9lNdCeV1A3YQAAAMBAI8EFAACoi29+Mw2FI8EFAACoiwsuSEPhSHABAABQFBJcAAAAFIUEFwAAAEXp5j+ZAQAAoAQjR/Y7gp6gH1w0oR9cABg8lNVAe/SDCwAAgFogwQUAAKiLo49OQ+FIcAEAAOrissvSUDgSXAAAABSFBBcAAABFIcEFAABAUegHFwAAoC5Gjep3BD1BP7hoQj+4ADB4KKuB9ugHFwAAALVAggsAAFAXn/1sGgpHG1wAAIC6uPrqfkfQE9TgAgAAoCgkuAAAACgKCS4AAACKQhtcAACAuth8835H0BP0g4sm9IMLAIOHshpoj35wAQAAUAskuAAAAHVxyCFpKBxtcAEAAOpi1qx+R9AT1OACAACgKCS4AAAAKAoJLgAAAIpCG1wAAIC6GD++3xH0BP3gogn94ALA4KGsBtqjH1wAAADUAgkuAABAXRxwQBoKRxtcAACAurj11n5H0BPU4AIAAKAoJLgAAAAoCgkuAAAAikIbXAAAgLqYMKHfEfQE/eCiCf3gAsDgoawG2qMfXAAAANQCCS4AAEBd7LdfGgpHG1wAAIC6mDev3xH0BDW4AAAAKAoJLgAAAIpCggsAAICi0AYXAACgLnbaqd8R9AT94KIJ/eACwOChrAbaox9cAAAA1AIJLgAAQF3stVcaCkcbXAAAgLpYuLDfEfQENbgAAAAoCgkuAAAAikKCCwAAgKLQBhcAAKAudt653xH0BP3gogn94ALA4KGsBtqjH1wAAADUAgkuAABAXeyySxoKRxtcAACAunjiiX5H0BPU4AIAAKAoJLgAAAAoCgkuAAAAikIbXAAAgLrYbbd+R9AT9IOLJvSDCwCDh7IaaI9+cAEAAFALJLgAAAB1MWlSGgpHggsAAICikOACAACgKCS4AAAAKAoJLgAAAIpCP7gAAAB1sc8+/Y6gJ+gHF03oBxcABg9lNdAe/eACAADU3eOPp6FwNFEAAACoi113TX9nzOhrGCsbNbgAAAAoCgkuAAAAikKCCwAAgKKQ4AIAAKAoPGQGAABQF1Om9DuCnqAfXDShH1wAGDyU1UB79IMLAABQdwsWpKFwNFEAAACoi733Tn/pBxcAAABYdZDgAgAAoCgkuAAAACgKCS4AAACKwkNmAAAAdXHwwf2OoCfoBxdN6AcXAAYPZTXQHv3gAgAA1N3cuWkoHE0UAAAA6mLy5PSXfnABAACAVQcJLgAAAIpCggsAAICikOACAACgKDxkBgAAUBeHHtrvCHqCfnDRhH5wAWDwUFYD7dEPLgAAQN3NmZOGwtFEAQAAoC4OPDD9pR9cAAAAYNVBggsAAICikOACAACgKCS4AAAAKAoPmQEAANTFUUf1O4KeoB9cNKEfXAAYPJTVQHv0gwsAAFB3s2aloXA0UQAAAKiLQw5Jf+kHFwAAAFh1kOACAACgKCS4AAAAKAoJLgAAAIrCQ2YAAAB18ZWv9DuCniDBBQAAqIvXvKbfEfQETRQAAADq4qqr0lA4anABAADq4sgj01/6wQUAAABWHSS4AAAAKAoJLgAAAIpCggsAAICi8JAZAABAXRx3XL8j6AkSXAAAgLqYMKHfEfQETRQAAADq4tJL01A4R0S/Y8AAsc0FAQADhrIaK8ykSelvIf3g2p4ZERNbx1ODCwAAgKKQ4AIAAKAoJLgAAAAoCgkuAAAAikI3YQAAAHVx0kn9jqAnSHABAADqYsst+x1BT9BEAQAAoC7OPz8NhaMfXDShH1wAGDyU1Vhh6AcXAAAAWPWQ4AIAAKAoJLgAAAAoCgkuAAAAikI3YQAAAHVx+un9jqAnSHABAADqYsyYfkfQEzRRAAAAqIuzz05D4egHF03oBxcABg9lNVYY+sEFAAAAVj0kuAAAACgKCS4AAACKQoILAACAotBNGAAAQF385Cf9jqAnSHDRZIcddtB1113X7zAAAMDKMHp0vyPoCZooAAAA1MW0aWkoHAkuAABAXZDgAgAAAKseElwAAAAUhQQXAAAARSHBBQAAQFHoJgwAAKAuLryw3xH0BAkuAABAXayzTr8j6AmaKAAAANTFCSekoXAkuAAAAHUxfXoaCkeCCwAAgKKQ4AIAAKAoJLgAAAAoCgkuAAAAiuKI6HcMGCC2H5E0p99xoMloSQv6HQT+Dudl8HBOBhPnZfCUdE7GRsTGrSPpBxet5kTExH4HgSVsX8c5GTycl8HDORlMnJfBU4dzQhMFAAAAFIUEFwAAAEUhwUWrk/sdAP4O52QwcV4GD+dkMHFeBk/x54SHzAAAAFAUanABAABQFBLcGrL9DttzbN9m+4g20237u3n6bNvb9yPOuunivOybz8ds21fZ3q4fcdbJcOekMt+Othfb3ruX8dVVN+fF9iTbs2zfZPuKXsdYN13cvzawfb7tG/I5+XA/4qwT2z+wfa/t33eYXnRZT4JbM7ZXl/SfknaRtLWkD9jeumW2XSRtkYcDJH2vp0HWUJfn5c+S3hgR20o6WjVoQ9VPXZ6Txnxfl3RRbyOsp27Oi+0NJZ0g6V0RsY2k9/Y6zjrp8rPyMUk3R8R2kiZJ+pbtNXsaaP1Mk/SOIaYXXdaT4NbPKyXdFhF/ioinJJ0laY+WefaQdFok10ja0PamvQ60ZoY9LxFxVUQ8kN9eI2nzHsdYN918ViTpXyT9VNK9vQyuxro5Lx+UdE5E/EWSIoJzs3J1c05C0nq2LWldSfdLWtTbMOslIn6pdJw7KbqsJ8Gtn80kza28n5fHLe08WLGW9ph/RNL/rtSIMOw5sb2ZpD0lndjDuOqum8/KeEkb2Z5he6btD/Usunrq5pwcL2krSX+VdKOkT0bEM70JDx0UXdbzn8zqx23GtXal0c08WLG6Pua236SU4L5upUaEbs7JcZI+ExGLU8UUeqCb8zJC0g6SdpY0UtLVtq+JiFtXdnA11c05ebukWZLeLOnFki6xfWVEPLySY0NnRZf1JLj1M0/SmMr7zZW+US/tPFixujrmtreVdKqkXSJiYY9iq6tuzslESWfl5Ha0pF1tL4qIn/Ukwnrq9h62ICIek/SY7V9K2k4SCe7K0c05+bCkr0Xqm/Q223+W9FJJv+1NiGij6LKeJgr1c62kLWy/MDfwf7+k81rmOU/Sh/ITlq+W9FBE3N3rQGtm2PNi+wWSzpE0mZqonhj2nETECyNiXESMk/QTSf9McrvSdXMP+7mk19seYXsdSa+SdEuP46yTbs7JX5Rq1GV7E0lbSvpTT6NEq6LLempwayYiFtn+uNIT36tL+kFE3GT7oDz9REkXStpV0m2SHlf65o2VqMvz8nlJoySdkGsMF0XExH7FXLouzwl6rJvzEhG32P6FpNmSnpF0akS07SoJy6/Lz8rRkqbZvlHpp/HPRMSCvgVdA7bPVOqxYrTteZK+IGkNqR5lPf/JDAAAAEWhiQIAAACKQoILAACAopDgAgAAoCgkuAAAACgKCS4AAACKQoILAGhie6rtw/LrV9v+je1Ztm+xPbVl3ufYXmh7g5bxP7O9zxDbeHSlBA8AIsEFAAztR5IOiIgJkl4maXp1Yv5vYRdLendjXE52Xyfpgp5FCQAVJLgAANn+N9tzbF+q9F+mGp4n6W5JiojFEXFzm8XPVPrvVQ17SvqFpNVsX2b7ets32t6jzXYn2b6g8v5421Py6x1sX2F7pu2LbG+ax3/C9s22Z9s+azl3HUCB+E9mAFBztndQSlBfoVQuXC9pZp58rKQ5tmcoJa0/iognW1bxC0mn2h4VEQvzuv5D0pOS9oyIh22PlnSN7fOii/8wZHuNvI49IuI+2++T9GVJ/yjpCEkvjIi/2d5wefYdQJmowQUAvF7SuRHxeEQ8rPQ/6iVJEfElSROVmiF8UCmZbRIRT+Vl9s6J7IQ8vyV9xfZsSZdK2kzSJl3GtKVSk4hLbM+SdJSkzfO02ZLOsL2fpEVLtacAaoEaXACAJHWsVY2I2yV9z/Ypku6r1NRWnamUhFrSzyPi6dzUYGNJO+T3d0hau2W5RWqubGlMt6SbImKnNiG9U9IbJL1L0udsbxMRJLoAnkUNLgDgl5L2tD3S9nqSdm9MsP1O285vt5C0WNKDbdbxf3n6x5SSXUnaQNK9Obl9k6SxbZa7U9LWttfKD6ftnMfPkbSx7Z1yHGvY3sb2apLGRMT/STpc0oaS1l3G/QZQKGpwAaDmIuJ622dLmqWUcF5ZmTxZ0rG2H1eqbd03Iha3Wccztn8q6b1KCbMknSHpfNvX5XX/oc1yc21PV2p28EdJv8vjn7K9t6Tv5sR3hKTjJN0q6cd5nCUdGxEPLtcBAFAcd9HWHwAAAFhl0EQBAAAARSHBBQAAQFFIcAEAAFAUElwAAAAUhQQXAAAARSHBBQAAQFFIcAEAAFAUElwAAAAU5f8BLIMyE4kqmUoAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 720x432 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# calculate dS values relative to a reference leaf\n",
    "# hard-code Paramecium tetraurelia because it is Queen\n",
    "\n",
    "def calculate_dS(reference):\n",
    "    partial_leaf_name = reference\n",
    "    output_txt = alnPrefix + '.ref.dS.txt'\n",
    "    output_fig  = alnPrefix + '.ref.dS.pdf'\n",
    "    # Initialize variables to store the reference leaf and its full name\n",
    "    reference_leaf = None\n",
    "    reference_leaf_name = None\n",
    "\n",
    "    # Search for the leaf with a name that starts with the specified partial name\n",
    "    for leaf in tree:\n",
    "        if leaf.name.startswith(partial_leaf_name):\n",
    "            reference_leaf = leaf\n",
    "            reference_leaf_name = leaf.name\n",
    "\n",
    "    if reference_leaf:\n",
    "        # Calculate dS values for leaves relative to the reference leaf\n",
    "        dS_values = {}\n",
    "\n",
    "        for leaf in tree:\n",
    "            # Extract TaxID from the leaf name\n",
    "            taxid = int(leaf.name.split('.')[0])\n",
    "\n",
    "            # Get the scientific name associated with the TaxID\n",
    "            species_name = ncbi.get_taxid_translator([taxid]).get(taxid, leaf.name)\n",
    "\n",
    "            # Calculate the dS value for the leaf relative to the reference leaf\n",
    "            dS = reference_leaf.get_distance(leaf)\n",
    "            dS_values[leaf.name] = {'Species': species_name, 'dS': dS}\n",
    "\n",
    "        with open(output_txt, \"w\") as f:\n",
    "            f.write('#gene\\tspecies\\tdS\\n')\n",
    "            for leaf_name, data in dS_values.items():\n",
    "                print(f\"{leaf_name}\\t{data['Species']}\\t{data['dS']:.4f}\")\n",
    "                f.write(f\"{leaf_name}\\t{data['Species']}\\t{data['dS']:.4f}\\n\")\n",
    "            print(f\"Ds values have been written to {output_txt}\")\n",
    "\n",
    "\n",
    "\n",
    "        # Create a bar chart\n",
    "        geneName = str(alnPrefix.split(\"_\")[0])\n",
    "        leaf_names = list(dS_values.keys())\n",
    "        species_names = [data['Species'] for data in dS_values.values()]\n",
    "        dS_values_list = [data['dS'] for data in dS_values.values()]\n",
    "\n",
    "        fig, ax = plt.subplots(figsize=(10, 6))\n",
    "        ax.barh(species_names, dS_values_list, color='black')\n",
    "        ax.set_xlabel('dS Values')\n",
    "        ax.set_title('dS Values relative to reference leaf: ' + partial_leaf_name + ' for gene ' + geneName)\n",
    "\n",
    "        # Set x-axis limit to 1.1 if dS values are less than 1.1\n",
    "        max_dS = max(dS_values_list)\n",
    "        if max_dS < 1.1:\n",
    "            ax.set_xlim(0, 1.1)\n",
    "\n",
    "        # Add a red dotted line at x=1\n",
    "        ax.axvline(x=1, color='red', linestyle='--')\n",
    "\n",
    "        # Save the plot\n",
    "        plt.savefig(output_fig)\n",
    "        plt.show()\n",
    "\n",
    "    else:\n",
    "        print(f\"No leaf found in the tree with a name starting with '{partial_leaf_name}'.\")\n",
    "\n",
    "calculate_dS('5888.PTET_51')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "d8b196e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# run the evolutionary models (longest step)\n",
    "# intelligently checking whether a model exists was a massive waste of time because ete3 is a fucking fuck and won't \"find\" the models after it is relaunched\n",
    "\n",
    "#def folder_exists(folderPath):\n",
    "#    return os.path.exists(folderPath) and os.path.isdir(folderPath)\n",
    "\n",
    "#model = [\"M1\", \"M2\", \"M7\", \"M8\", \"M0\", \"fb\"]\n",
    "models = ['M1', 'M2', 'M7', 'M8']\n",
    "\n",
    "#for model in models:\n",
    "#    if not folder_exists(model):\n",
    "#        tree.run_model(model)\n",
    "#    else:\n",
    "#        print(f\"The model {model} exists. Skipping\")\n",
    "for model in models:\n",
    "    print(model)\n",
    "    tree.run_model(model)\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "b387a2fb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.008258931980424156\n",
      "M8 model wins.\n"
     ]
    }
   ],
   "source": [
    "def get_pvals(alt, neg):\n",
    "\n",
    "    altModel = tree.get_evol_model(alt)\n",
    "    pval = tree.get_most_likely(alt, neg)\n",
    "    print(pval)\n",
    "    output_file = alnPrefix + '.pvals.txt'\n",
    "    with open(output_file, \"a\") as output:\n",
    "        output.write('altModel\\tnegModel\\tpval\\n')\n",
    "        output.write(alt + '\\t' + neg  +  '\\t' + str(pval) + '\\n')\n",
    "\n",
    "    if pval < 0.05:\n",
    "        output_file2 = alnPrefix + '.sigAAs.txt'\n",
    "        with open(output_file2, \"a\") as output2:\n",
    "            output2.write('alt_model\\tnegmodel\\taa\\tposition\\tprobability\\n')\n",
    "            print(f'{alt} model wins.')\n",
    "            for s in range(len(altModel.sites['BEB']['aa'])):\n",
    "                    if altModel.sites['BEB']['p10'][s] > 0.95:\n",
    "                        pos_acid = altModel.sites['BEB']['aa'][s]\n",
    "                        pos_site = str(s)\n",
    "                        pos_prob = str(altModel.sites['BEB']['p10'][s])\n",
    "                        pos_omega = str(altModel.sites['BEB']['w'][s])\n",
    "                        p0 = altModel.sites['BEB']['p0'][s]\n",
    "                        p1 = altModel.sites['BEB']['p1'][s]\n",
    "                        p2 = altModel.sites['BEB']['p2'][s]\n",
    "                        p3 = altModel.sites['BEB']['p3'][s]\n",
    "                        p4 = altModel.sites['BEB']['p4'][s]\n",
    "                        p5 = altModel.sites['BEB']['p5'][s]\n",
    "                        p6 = altModel.sites['BEB']['p6'][s]\n",
    "                        p7 = altModel.sites['BEB']['p7'][s]\n",
    "                        p8 = altModel.sites['BEB']['p8'][s]\n",
    "                        p9 = altModel.sites['BEB']['p9'][s]\n",
    "                        p10 = altModel.sites['BEB']['p10'][s]\n",
    "                        pv = altModel.sites['BEB']['pv'][s]\n",
    "                        w = altModel.sites['BEB']['w'][s]\n",
    "                        se = altModel.sites['BEB']['se'][s]\n",
    "                        \n",
    "                        #print(s, p0, p1, p2, p3, p4, p5, p6, p7, p8, \"p9:\", p9, \"p10:\", p10, \"pv:\", pv, w, se)\n",
    "                        \n",
    "                        output_line2 = alt + '\\t' + neg + '\\t' + pos_acid + '\\t' + pos_site + '\\t' + pos_prob + '\\t' + pos_omega + '\\t' + '\\n'\n",
    "                        print(output_line2)\n",
    "                        #print(altModel.sites['BEB'].keys())\n",
    "                        output2.write(output_line2)\n",
    "                        #print(altModel.sites['BEB']['p8'][s])\n",
    "    else:\n",
    "        print(f'{neg} model is not rejected')\n",
    "\n",
    "        \n",
    "get_pvals('M8', 'M7')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "1511556b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "M1\n",
      " Evolutionary Model M1:\n",
      "        log likelihood       : -2665.441654\n",
      "        number of parameters : 8\n",
      "        sites inference      : NEB\n",
      "        sites classes        : \n",
      "        proportions : p0=0.96733   p1=0.03267   \n",
      "        w           : w0=0.04988   w1=1.0       \n",
      "        branches             : \n",
      "        mark:     , omega: None      , node_ids: 5   , name: Inner2\n",
      "        mark:     , omega: 0.0809    , node_ids: 6   , name: Inner1\n",
      "        mark:     , omega: 0.0809    , node_ids: 2   , name: 5888.PTET_51_T0360256\n",
      "        mark:     , omega: 0.0809    , node_ids: 1   , name: 5888.PTET_51_T0210034\n",
      "        mark:     , omega: 0.0809    , node_ids: 3   , name: 5888.PTET_51_T0460249\n",
      "        mark:     , omega: 0.0809    , node_ids: 4   , name: 5888.PTET_51_T1750015\n",
      "        \n",
      "M2\n",
      " Evolutionary Model M2:\n",
      "        log likelihood       : -2664.867291\n",
      "        number of parameters : 10\n",
      "        sites inference      : BEB, NEB\n",
      "        sites classes        : \n",
      "        proportions : p0=0.97733   p1=0.0       p2=0.02267   \n",
      "        w           : w0=0.05276   w1=1.0       w2=1.82314   \n",
      "        branches             : \n",
      "        mark:     , omega: None      , node_ids: 5   , name: Inner2\n",
      "        mark:     , omega: 0.0929    , node_ids: 6   , name: Inner1\n",
      "        mark:     , omega: 0.0929    , node_ids: 2   , name: 5888.PTET_51_T0360256\n",
      "        mark:     , omega: 0.0929    , node_ids: 1   , name: 5888.PTET_51_T0210034\n",
      "        mark:     , omega: 0.0929    , node_ids: 3   , name: 5888.PTET_51_T0460249\n",
      "        mark:     , omega: 0.0929    , node_ids: 4   , name: 5888.PTET_51_T1750015\n",
      "        \n",
      "M7\n",
      " Evolutionary Model M7:\n",
      "        log likelihood       : -2670.624256\n",
      "        number of parameters : 8\n",
      "        sites inference      : NEB\n",
      "        sites classes        : \n",
      "        proportions : p0=0.125     p1=0.125     p2=0.125     p3=0.125     p4=0.125     p5=0.125     p6=0.125     p7=0.125     \n",
      "        w           : w0=0.02807   w1=0.04085   w2=0.05041   w3=0.05939   w4=0.06879   w5=0.07964   w6=0.09395   w7=0.12014   \n",
      "        branches             : \n",
      "        mark:     , omega: None      , node_ids: 5   , name: Inner2\n",
      "        mark:     , omega: 0.0677    , node_ids: 6   , name: Inner1\n",
      "        mark:     , omega: 0.0677    , node_ids: 2   , name: 5888.PTET_51_T0360256\n",
      "        mark:     , omega: 0.0677    , node_ids: 1   , name: 5888.PTET_51_T0210034\n",
      "        mark:     , omega: 0.0677    , node_ids: 3   , name: 5888.PTET_51_T0460249\n",
      "        mark:     , omega: 0.0677    , node_ids: 4   , name: 5888.PTET_51_T1750015\n",
      "        \n",
      "M8\n",
      " Evolutionary Model M8:\n",
      "        log likelihood       : -2665.827796\n",
      "        number of parameters : 10\n",
      "        sites inference      : BEB, NEB\n",
      "        sites classes        : \n",
      "        proportions : p0=0.1225    p1=0.1225    p2=0.1225    p3=0.1225    p4=0.1225    p5=0.1225    p6=0.1225    p7=0.1225    p8=0.02001   \n",
      "        w           : w0=0.02857   w1=0.03888   w2=0.04633   w3=0.05319   w4=0.06027   w5=0.06833   w6=0.07885   w7=0.09786   w8=1.95658   \n",
      "        branches             : \n",
      "        mark:     , omega: None      , node_ids: 5   , name: Inner2\n",
      "        mark:     , omega: 0.097     , node_ids: 6   , name: Inner1\n",
      "        mark:     , omega: 0.097     , node_ids: 2   , name: 5888.PTET_51_T0360256\n",
      "        mark:     , omega: 0.097     , node_ids: 1   , name: 5888.PTET_51_T0210034\n",
      "        mark:     , omega: 0.097     , node_ids: 3   , name: 5888.PTET_51_T0460249\n",
      "        mark:     , omega: 0.097     , node_ids: 4   , name: 5888.PTET_51_T1750015\n",
      "        \n"
     ]
    }
   ],
   "source": [
    "# TABARNAK WHY THE FUCK CANT THIS STUPID FUCK ETE3 FIND THE FUCKING MODELS IN THE FUCKING WORKING DIRECTORY I FUCKING SPECIFIED\n",
    "\n",
    "output_models = alnPrefix + '.models'\n",
    "with open(output_models, \"w\") as output:\n",
    "    for model in models:\n",
    "        print(model)\n",
    "        get_model = tree.get_evol_model(model)\n",
    "        print(get_model)\n",
    "        output.write(str(get_model)+'\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "8ef0b85c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "M1 model is not rejected\n",
      "M8 model wins.\n",
      "positively selected site 48 at position: 48 with probability: 0.12965\n",
      "positively selected site 50 at position: 50 with probability: 0.10712\n",
      "positively selected site 51 at position: 51 with probability: 0.09468\n",
      "positively selected site 76 at position: 76 with probability: 0.12677\n",
      "positively selected site 82 at position: 82 with probability: 0.06181\n",
      "positively selected site 101 at position: 101 with probability: 0.12116\n",
      "positively selected site 104 at position: 104 with probability: 0.09389\n",
      "positively selected site 128 at position: 128 with probability: 0.05086\n",
      "positively selected site 181 at position: 181 with probability: 0.13579\n",
      "positively selected site 231 at position: 231 with probability: 0.10455\n",
      "positively selected site 238 at position: 238 with probability: 0.12402\n",
      "positively selected site 253 at position: 253 with probability: 0.06887\n",
      "positively selected site 273 at position: 273 with probability: 0.05395\n",
      "positively selected site 274 at position: 274 with probability: 0.15134\n",
      "positively selected site 276 at position: 276 with probability: 0.11087\n",
      "positively selected site 283 at position: 283 with probability: 0.0508\n",
      "positively selected site 284 at position: 284 with probability: 0.09384\n",
      "positively selected site 287 at position: 287 with probability: 0.0607\n",
      "positively selected site 289 at position: 289 with probability: 0.11325\n",
      "positively selected site 296 at position: 296 with probability: 0.07629\n",
      "positively selected site 308 at position: 308 with probability: 0.10049\n"
     ]
    }
   ],
   "source": [
    "     "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "6a7ab683",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "def evol_graphs(alt, neg, suffix):\n",
    "    \n",
    "    # Change the colours\n",
    "    col_line = {'NS' : '#000000', \n",
    "           'RX' : '#000000', 'RX+': '#000000', \n",
    "           'CN' : '#000000', 'CN+': '#000000', \n",
    "           'PS' : '#000000', 'PS+': '#000000'}\n",
    "    col_bar = {'NS' : '#BCBCBC', \n",
    "           'RX' : '#5D63AB', 'RX+': '#5D63AB', \n",
    "           'CN' : '#659A62', 'CN+': '#659A62', \n",
    "           'PS' : '#F4C95D', 'PS+': '#F4C95D'}\n",
    "\n",
    "    # Print the first tree without naming branches\n",
    "    alt_model = tree.get_evol_model(alt)\n",
    "    neg_model = tree.get_evol_model(neg)\n",
    "\n",
    "    alt_model.set_histface(up=True, colors=col_line, errors=True, kind='curve', ylim=[0,20], hlines = [1], hlines_col=['black'])\n",
    "    tree.render(alnPrefix +  '.' + alt + suffix + '.line.pdf', histfaces=[alt])\n",
    "\n",
    "    alt_model.set_histface(up=True, colors=col_bar, errors=True, kind='stick', ylim=[0,4], hlines = [1], hlines_col=['black'])\n",
    "    tree.render(alnPrefix +  '.' + alt + suffix + '.bar.pdf', histfaces=[alt])\n",
    "\n",
    "def modify_leaf_names_reroot():\n",
    "    # Modify the tree leaf names\n",
    "    # ONLY DO THIS AS THE LAST STEP\n",
    "    for leaf in tree.iter_leaves():\n",
    "        taxid = int(leaf.name.split('.')[0])\n",
    "        species_name = get_species_name(taxid)\n",
    "        leaf.name = species_name.split(' ')[1]\n",
    "        \n",
    "    preference_dictionary = {'tet':1, \n",
    "                         'qua':2, 'pen':2, 'pri':2, 'nov':2, 'dod':2, 'oct':2, 'sex':2, 'son':2, 'tre': 2,\n",
    "                         'cau':3}\n",
    "    root_point = tree.get_farthest_oldest_leaf(preference_dictionary)\n",
    "    tree.set_outgroup(root_point)\n",
    "\n",
    "    \n",
    "\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "914f3485",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'modify_leaf_names' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m-------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                               Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-14-1f5672e8e3c5>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mevol_graphs\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'M2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'M1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m''\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0mevol_graphs\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'M8'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'M7'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m''\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0mmodify_leaf_names\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m \u001b[0mevol_graphs\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'M2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'M1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'_sp'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0mevol_graphs\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'M8'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'M7'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'_sp'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'modify_leaf_names' is not defined"
     ]
    }
   ],
   "source": [
    "\n",
    "evol_graphs('M2', 'M1', '')\n",
    "evol_graphs('M8', 'M7', '')\n",
    "modify_leaf_names()\n",
    "evol_graphs('M2', 'M1', '_sp')\n",
    "evol_graphs('M8', 'M7', '_sp')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "118c6c92",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "EvolTree running in new folder: /Users/jra/Dropbox/ciliates/P0460249_riboprotein/P0460249_riboprotein.fa.tblastn.fa.trimal.automated1.extract.evolTree\n",
      "\n",
      "      /-224956.PQUADEC_NiA_T00270028\n",
      "     |\n",
      "     |      /-43138.PPENT_87_T0670067\n",
      "   /-|   /-|\n",
      "  |  |  |  |   /-5886.PPRIM_AZ9_T0580068\n",
      "  |  |  |   \\-|\n",
      "  |   \\-|      \\-5886.PPRIM_Ir42_T15418\n",
      "  |     |\n",
      "  |     |   /-219701.PNOV_TE_T00080084\n",
      "  |     |  |\n",
      "  |      \\-|   /-304694.PDODEC_274_T01140047\n",
      "  |        |  |\n",
      "  |         \\-|      /-5888.PTET_51_T0460249\n",
      "  |           |   /-|\n",
      "--|           |  |   \\-412030.PTET_42_T00015928001\n",
      "  |            \\-|\n",
      "  |              |   /-43137.POCT_138_T0740045\n",
      "  |               \\-|\n",
      "  |                  \\-43137.POCT_K8_T71800002770320265\n",
      "  |\n",
      "  |   /-5885.PCAU_43c3d_T00080186\n",
      "  |--|\n",
      "  |  |   /-65128.PSEX_AZ84_T15940\n",
      "  |   \\-|\n",
      "  |      \\-65129.PSON_ATCC_T0490080\n",
      "  |\n",
      "   \\-65130.PTRED_209_T71800001293390218\n",
      "\n",
      "   /-caudatum\n",
      "  |\n",
      "  |      /-sexaurelia\n",
      "  |   /-|\n",
      "  |  |   \\-sonneborni\n",
      "--|  |\n",
      "  |  |      /-quadecaurelia\n",
      "  |  |     |\n",
      "  |  |     |      /-pentaurelia\n",
      "  |  |   /-|   /-|\n",
      "   \\-|  |  |  |  |   /-primaurelia\n",
      "     |  |  |  |   \\-|\n",
      "     |  |   \\-|      \\-primaurelia\n",
      "     |  |     |\n",
      "     |  |     |   /-novaurelia\n",
      "     |  |     |  |\n",
      "     |  |      \\-|   /-dodecaurelia\n",
      "      \\-|        |  |\n",
      "        |         \\-|      /-tetraurelia\n",
      "        |           |   /-|\n",
      "        |           |  |   \\-tetraurelia\n",
      "        |            \\-|\n",
      "        |              |   /-octaurelia\n",
      "        |               \\-|\n",
      "        |                  \\-octaurelia\n",
      "        |\n",
      "         \\-tredecaurelia\n"
     ]
    }
   ],
   "source": [
    "tree, alnPrefix = organize_your_life('/Users/jra/Dropbox/ciliates/P0460249_riboprotein/P0460249_riboprotein.fa.tblastn.fa.trimal.automated1.extract')\n",
    "print(tree)\n",
    "modify_leaf_names_reroot()\n",
    "\n",
    "print(tree)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "3a40418c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " Evolutionary Model fb:\n",
      "        log likelihood       : -2669.541875\n",
      "        number of parameters : 11\n",
      "        sites inference      : None\n",
      "        sites classes        : None\n",
      "        branches             : \n",
      "        mark: #0  , omega: None      , node_ids: 5   , name: Inner2\n",
      "        mark: #1  , omega: 0.0497    , node_ids: 6   , name: Inner1\n",
      "        mark: #2  , omega: 0.5455    , node_ids: 2   , name: 5888.PTET_51_T0360256\n",
      "        mark: #3  , omega: 507.2739  , node_ids: 1   , name: 5888.PTET_51_T0210034\n",
      "        mark: #4  , omega: 0.0336    , node_ids: 3   , name: 5888.PTET_51_T0460249\n",
      "        mark: #5  , omega: 0.0956    , node_ids: 4   , name: 5888.PTET_51_T1750015\n",
      "        \n",
      "None\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/jra/miniconda3/envs/ete3/lib/python3.6/site-packages/ete3/evol/evoltree.py:459: UserWarning: Model  Evolutionary Model fb:\n",
      "        log likelihood       : -2669.541875\n",
      "        number of parameters : 11\n",
      "        sites inference      : None\n",
      "        sites classes        : None\n",
      "        branches             : \n",
      "        mark: #0  , omega: None      , node_ids: 5   , name: Inner2\n",
      "        mark: #1  , omega: 0.0497    , node_ids: 6   , name: Inner1\n",
      "        mark: #2  , omega: 0.5455    , node_ids: 2   , name: 5888.PTET_51_T0360256\n",
      "        mark: #3  , omega: 507.2739  , node_ids: 1   , name: 5888.PTET_51_T0210034\n",
      "        mark: #4  , omega: 0.0336    , node_ids: 3   , name: 5888.PTET_51_T0460249\n",
      "        mark: #5  , omega: 0.0956    , node_ids: 4   , name: 5888.PTET_51_T1750015\n",
      "         not found.\n",
      "  warn(\"Model %s not found.\" % (modelname))\n"
     ]
    }
   ],
   "source": [
    "\n",
    "tree.run_model('fb')\n",
    "fb_results = tree.get_evol_model('fb')\n",
    "print(fb_results)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7d64ee5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# https://python.hotexamples.com/examples/ete3/EvolTree/run_model/python-evoltree-run_model-method-examples.html\n",
    "\n",
    "\n",
    "from ete3 import TreeStyle\n",
    "from ete3 import EvolTree\n",
    "from ete3 import faces\n",
    "\n",
    "tree = EvolTree(\"data/S_example/measuring_S_tree.nw\")\n",
    "tree.link_to_alignment('data/S_example/alignment_S_measuring_evol.fasta')\n",
    "\n",
    "print(tree)\n",
    "\n",
    "print('\\n Running free-ratio model with calculation of ancestral sequences...')\n",
    "\n",
    "tree.run_model('fb_anc')\n",
    "#tree.link_to_evol_model('/tmp/ete3-codeml/fb_anc/out', 'fb_anc')\n",
    "\n",
    "I = TreeStyle()\n",
    "I.force_topology = False\n",
    "I.draw_aligned_faces_as_table = True\n",
    "I.draw_guiding_lines = True\n",
    "I.guiding_lines_type = 2\n",
    "I.guiding_lines_color = \"#CCCCCC\"\n",
    "for n in sorted(tree.get_descendants() + [tree], key=lambda x: x.node_id):\n",
    "    if n.is_leaf(): continue\n",
    "    anc_face = faces.SequenceFace(n.sequence, 'aa', fsize=10, bg_colors={})\n",
    "    I.aligned_foot.add_face(anc_face, 1)\n",
    "    I.aligned_foot.add_face(\n",
    "        faces.TextFace('node_id: #%d ' % (n.node_id), fsize=8), 0)\n",
    "print('display result of bs_anc model, with ancestral amino acid sequences.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2b505bc",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
